core/src/main/scala/kafka/log/LogCleanerManager.scala

```diff
@@ -240,3 +236,66 @@ private[log] class LogCleanerManager(val logDirs: Array[File], val logs: Pool[To
     }
   }
 }
+
+private[log] object LogCleanerManager extends Logging {
+
+  /**
+    * Returns the range of dirty offsets that can be cleaned.
+    *
+    * @param log the log
+    * @param lastClean the map of checkpointed offsets
+    * @param now the current time in milliseconds of the cleaning operation
+    * @return the lower (inclusive) and upper (exclusive) offsets
+    */
+  def cleanableOffsets(log: Log, topicAndPartition: TopicAndPartition, lastClean: immutable.Map[TopicAndPartition, Long], now: Long): (Long, Long) = {
+
+    // the checkpointed offset, ie., the first offset of the next dirty segment
+     val lastCleanOffset: Option[Long] = lastClean.get(topicAndPartition)
+
+    // If the log segments were truncated, the checkpointed offset is no longer valid;
+    // reset to the log starting offset and log the error
+    val logStartOffset = log.logSegments.head.baseOffset
+    val firstDirtyOffset = {
+      val offset = lastCleanOffset.getOrElse(logStartOffset)
+      if (offset < logStartOffset) {
+        error("Resetting first dirty offset to log start offset %d since the checkpointed offset %d is invalid."
+          .format(logStartOffset, offset))
+        logStartOffset
+      } else {
+        offset
+      }
+    }
+
+    // dirty log segments
+    val dirtySegments = log.logSegments(firstDirtyOffset, log.activeSegment.baseOffset).toArray :+ log.activeSegment
+
+    val compactionLagMs = math.max(log.config.compactionLagMs, 0L)
+
+    // find first segment that cannot be cleaned
+    // neither the active segment, nor segments with any messages closer to the head of the log than the minimum compaction lag time
+    // may be cleaned
+    val firstUncleanableDirtyOffset = Seq (
+
+        // the active segment is always uncleanable
+        Option(log.activeSegment.baseOffset),
+
+        // the first segment whose last message time is within a minimum time lag from now
+        if (compactionLagMs > 0) {
+          dirtySegments.find {
+            case segment =>
+              segment.log.lastOption match {
```

Comment:
Since KAFKA-3163 is still not merged can we go back to using the (less accurate but efficient) log segment modification time? Once the time indices are available compaction could be modified to use them in a subsequent PR.